TrackletAlgorithm/VMRouterCMTop.cc

```diff
@@ -0,0 +1,169 @@
+#include "VMRouterCMTop.h"
+
+// VMRouter Top Function for layer 2, AllStub region A
+// Sort stubs into smaller regions in phi, i.e. Virtual Modules (VMs).
+
+// NOTE: to run a different phi region, change the following
+//          - constants specified in VMRouterCMTop.h
+//          - add/remove pragmas depending on inputStubs in VMRouterCMTop.cc
+//          - maskASI in VMRouterCMTop.cc
+//          - add the phi region in emData/download.sh, make sure to also run clean
+
+
+void VMRouterCMTop(const BXType bx, BXType& bx_o
+	// Input memories
+	, const InputStubMemory<inputType> inputStubs[numInputs]
+#if kDISK > 0
+  , const InputStubMemory<DISK2S> inputStubsDisk2S[numInputsDisk2S]
+#endif
+
+	// Output memories
+	, AllStubMemory<outputType> memoriesAS[numASCopies]
+#if kLAYER == 1 || kLAYER == 2 || kLAYER == 3 || kLAYER == 5 || kDISK == 1 || kDISK == 3
+	, AllStubInnerMemory<outputType> memoriesASInner[numASInnerCopies]
+#endif
+	, VMStubMEMemoryCM<outputType, rzSizeME, phiRegSize> *memoryME
+#if kLAYER == 2 || kLAYER == 3 || kLAYER == 4 || kLAYER == 6 || kDISK == 1 || kDISK == 2 || kDISK == 4
+	, VMStubTEOuterMemoryCM<outputType,rzSizeTE,phiRegSize,numTEOCopies> *memoryTEO
+#endif
+	) {
+
+	// // Takes 2 clock cycles before on gets data, used at high frequencies
+	#pragma HLS resource variable=inputStubs[0].get_mem() latency=2
+	#pragma HLS resource variable=inputStubs[1].get_mem() latency=2
+	// #pragma HLS resource variable=inputStubs[2].get_mem() latency=2
+	// #pragma HLS resource variable=inputStubs[3].get_mem() latency=2
+
+	// Masks of which memories that are being used. The first memory is represented by the LSB
+	// and a "1" implies that the specified memory is used for this phi region
+	// First three bits (LSB) are the six A-F for Barrel, then the three after that are L,M,R for Barrel and disk, last three are L,M,R for Overlap
+	// NOTE: read from right to left (OR, OM, OL, BR/DR, BM/DM, BL/DL, BF, BE, BD, BC, BB, BA)
+	static const ap_uint<maskASIsize> maskASI = 0b110110000000; // Change me when switching phi region
```

Comment:
As with the parameters in VMRouterCMTop.h, would it make sense to also define this as a `constexpr` function, since you indicate that it depends on the phi region?

Reply:
I see you moved this to the header file, but I think the question is still relevant.